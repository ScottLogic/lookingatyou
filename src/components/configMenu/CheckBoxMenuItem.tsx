/* tslint:disable:jsx-no-lambda */

import React from 'react';
export default function CheckBoxMenuItem(props: {
    name: string;
    onInputChange: (checked: boolean) => void;
    checked: boolean;
}) {
    return (
        <div>
            <label>{props.name}</label>
            <input
                type="checkbox"
                checked={props.checked}
                onChange={event => {
                    props.onInputChange(event.target.checked);
                }}
            />
        </div>
    );
}
