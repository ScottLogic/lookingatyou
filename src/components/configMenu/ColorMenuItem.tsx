/* tslint:disable: jsx-no-lambda */

import React from 'react';
export default function TextBoxMenuItem(props: {
    name: string;
    onInputChange: (text: string) => void;
    color: string;
}) {
    return (
        <div>
            <label>{props.name}</label>
            <input
                type="color"
                value={props.color}
                onChange={event => {
                    props.onInputChange(event.target.value);
                }}
            />
        </div>
    );
}
