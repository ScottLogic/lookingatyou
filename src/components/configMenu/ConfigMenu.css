:root {
    --main-background-colour: #222;
    --menu-background-colour: #111;
    --menu-text-colour: #818181;
    --menu-top-padding: 60px;
    --menu-z-index: 10;
    --canvas-z-index: 20;
    --canvas-width: 100%;
}

.ConfigMenu {
    /* Position */
    top: 0;
    left: 0;
    height: calc(100% - var(--menu-top-padding));
    width: 0;
    position: fixed;

    /* View */
    padding-top: var(--menu-top-padding);
    background-color: var(--menu-background-colour);
    overflow-x: hidden;
    z-index: var(--menu-z-index);
    display: inline-block;

    /* Font */
    text-decoration: none;
    color: var(--menu-text-colour);
    font-size: 1.2em;

    transition: 500ms;
}

.ConfigMenu a,
.ConfigMenu label,
.ConfigMenu input,
.ConfigMenu select,
.ConfigMenu h1 {
    padding: 10px;
    width: calc(var(--canvas-width) / 2);
    display: inline-block;
}

.ConfigMenu input {
    width: calc(var(--canvas-width) / 4);
}

.ConfigMenu select {
    width: calc(var(--canvas-width) / 2.9);
}

.ConfigMenu #canvas {
    width: var(--canvas-width);
}

.icon {
    width: 2em;
    border-radius: 25%;
    color: #fff;
    line-height: 2em;
    text-align: center;
    background: rgb(61, 61, 61);
    display: inline-block;
    top: -0.25em;
}
