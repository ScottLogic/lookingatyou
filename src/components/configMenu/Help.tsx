import React, { Fragment } from 'react';
import ReactTooltip from 'react-tooltip';

export enum HelpWith {
    FPS,
    X_SENSITIVITY,
    Y_SENSITIVITY,
    VIDEO_STREAM,
    SWAP_EYES,
    IRIS_COLOUR,
    APP,
}

const text = (section: HelpWith) => {
    switch (section) {
        case HelpWith.FPS:
            return (
                <Fragment>
                    This number sets how often the eyes will update in Frames
                    per Second.
                </Fragment>
            );
        case HelpWith.X_SENSITIVITY:
            return (
                <Fragment>
                    This number sets how sensitive the eyes will be from left to
                    right.
                </Fragment>
            );
        case HelpWith.Y_SENSITIVITY:
            return (
                <Fragment>
                    This number sets how sensitive the eyes will be from top to
                    bottom.
                </Fragment>
            );
        case HelpWith.VIDEO_STREAM:
            return (
                <Fragment>
                    This shows you the current webcam feed, and outlines the
                    person that is currently being tracked.
                </Fragment>
            );
        case HelpWith.SWAP_EYES:
            return (
                <Fragment>
                    This allows you to swap the eyes when using two cameras so
                    you can correctly set the left and the right.
                </Fragment>
            );
        case HelpWith.IRIS_COLOUR:
            return (
                <Fragment>
                    This button allows you to select the colour of the eyes.
                </Fragment>
            );
        case HelpWith.APP:
            return (
                <Fragment>
                    This is an app designed to track users using a webcam. The
                    eyes will follow you around and react differently depending
                    on what they see. Hover over any of the menu items for more
                    information.
                </Fragment>
            );
        default:
            return '';
    }
};

interface IHelpProps {
    problemWith: HelpWith;
}

function Help(props: IHelpProps) {
    return (
        <div className={HelpWith[props.problemWith]}>
            <ReactTooltip
                id={HelpWith[props.problemWith]}
                place="right"
                type="dark"
                effect="solid"
            >
                <span>{text(props.problemWith)}</span>
            </ReactTooltip>
        </div>
    );
}

export default Help;
