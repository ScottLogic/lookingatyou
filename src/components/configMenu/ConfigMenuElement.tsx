import React from 'react';
import { connect } from 'react-redux';
import { updateConfigAction } from '../../store/actions/config/actions';
import { IRootStore } from '../../store/reducers/rootReducer';
import { getConfig } from '../../store/selectors/configSelectors';
import store from '../../store/store';
import ConfigMenu from './ConfigMenu';
import IUserConfig from './IUserConfig';
import CanvasMenuItem from './menuItems/CanvasMenuItem';
import CheckBoxMenuItem from './menuItems/CheckBoxMenuItem';
import ColorMenuItem from './menuItems/ColorMenuItem';
import TextBoxMenuItem from './menuItems/TextBoxMenuItem';

interface IConfigMenuElementProps {
    storage: Storage;
    config: IUserConfig;
    window: Window;
}

interface IConfigMenuElementMapStateToProps {
    config: IUserConfig;
}

export type ConfigMenuElementProps = IConfigMenuElementProps &
    IConfigMenuElementMapStateToProps;

const mapStateToProps = (
    state: IRootStore,
): IConfigMenuElementMapStateToProps => {
    return {
        config: getConfig(state),
    };
};

export function ConfigMenuElement(props: ConfigMenuElementProps) {
    function parseAndStoreXSensitivity(xSensitivity: string) {
        store.dispatch(
            updateConfigAction({
                partialConfig: {
                    xSensitivity: parseFloat(xSensitivity),
                },
            }),
        );
    }

    function parseAndStoreYSensitivity(ySensitivity: string) {
        store.dispatch(
            updateConfigAction({
                partialConfig: {
                    ySensitivity: parseFloat(ySensitivity),
                },
            }),
        );
    }

    function parseAndStoreFPS(fps: string) {
        store.dispatch(
            updateConfigAction({
                partialConfig: { fps: parseInt(fps, 10) },
            }),
        );
    }

    function storeSwapEyes(swapEyes: boolean) {
        store.dispatch(updateConfigAction({ partialConfig: { swapEyes } }));
    }

    function storeToggleDebug(toggleDebug: boolean) {
        store.dispatch(
            updateConfigAction({
                partialConfig: { toggleDebug },
            }),
        );
    }

    function storeIrisColor(irisColor: string) {
        store.dispatch(updateConfigAction({ partialConfig: { irisColor } }));
    }

    function extractFloatToString(floatString: string): string {
        return `${parseFloat(floatString)}`;
    }

    function extractIntToString(intString: string): string {
        return `${parseInt(intString, 10)}`;
    }

    function isValidSensitivity(sensitivity: string): boolean {
        return !isNaN(parseFloat(sensitivity)) && parseFloat(sensitivity) >= 0;
    }

    function isValidFPS(fps: string): boolean {
        return !isNaN(parseInt(fps, 10)) && parseInt(fps, 10) > 0;
    }

    return (
        <ConfigMenu width="14em" timerLength={1000} window={props.window}>
            <TextBoxMenuItem
                name={'X Sensitivity'}
                defaultValue={`${props.config.xSensitivity}`}
                isValidInput={isValidSensitivity}
                onValidInput={parseAndStoreXSensitivity}
                parse={extractFloatToString}
            />
            <TextBoxMenuItem
                name={'Y Sensitivity'}
                defaultValue={`${props.config.ySensitivity}`}
                isValidInput={isValidSensitivity}
                onValidInput={parseAndStoreYSensitivity}
                parse={extractFloatToString}
            />
            <TextBoxMenuItem
                name={'FPS'}
                defaultValue={`${props.config.fps}`}
                isValidInput={isValidFPS}
                onValidInput={parseAndStoreFPS}
                parse={extractIntToString}
            />
            <CheckBoxMenuItem
                name={'Swap Eyes'}
                checked={props.config.swapEyes}
                onInputChange={storeSwapEyes}
            />
            <CheckBoxMenuItem
                name={'Toggle Debug'}
                checked={props.config.toggleDebug}
                onInputChange={storeToggleDebug}
            />
            <ColorMenuItem
                name={'Iris Color'}
                color={props.config.irisColor}
                onInputChange={storeIrisColor}
            />
            <CanvasMenuItem name={'Left Camera'} />
            <CanvasMenuItem name={'Right Camera'} />
        </ConfigMenu>
    );
}

export default connect(mapStateToProps)(ConfigMenuElement);
