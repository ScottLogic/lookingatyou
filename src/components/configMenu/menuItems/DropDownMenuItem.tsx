import { FormControl, MenuItem, Select } from '@material-ui/core';
import React from 'react';
import { UpdateConfigAction } from '../../../store/actions/config/types';
import { HelpWith } from '../Help';

export interface IDropDownMenuItemProps {
    name: string;
    configName: string;
    onInputChange: UpdateConfigAction;
    values: string[];
    defaultValue: string;
    helpWith: HelpWith;
}

const DropDownMenuItem = React.memo(
    (props: IDropDownMenuItemProps) => {
        function getMenuItems(): JSX.Element[] {
            const menuItem = props.values.map((element, index) => (
                <MenuItem key={index} value={element}>
                    {element}
                </MenuItem>
            ));
            return menuItem;
        }

        function handleChange(event: any) {
            props.onInputChange({
                [props.configName]: event.target.value,
            });
        }

        return (
            <div data-tip={true} data-for={HelpWith[props.helpWith]}>
                <label>{props.name}</label>
                <FormControl className="selectForm">
                    <Select
                        value={props.defaultValue}
                        onChange={handleChange}
                        name={props.name}
                    >
                        {getMenuItems()}
                    </Select>
                </FormControl>
            </div>
        );
    },
    (previous, next) =>
        previous.defaultValue === next.defaultValue &&
        previous.values.length === next.values.length &&
        previous.values.every(
            (element, index) => element === next.values[index],
        ),
);

export default DropDownMenuItem;
