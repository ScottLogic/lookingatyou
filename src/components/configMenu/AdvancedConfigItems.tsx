import React from 'react';
import { IConfigState, PartialConfig } from '../../store/actions/config/types';
import { HelpWith } from './Help';
import CanvasMenuItem from './menuItems/CanvasMenuItem';
import CheckBoxMenuItem from './menuItems/CheckBoxMenuItem';
import DropDownMenuItem from './menuItems/DropDownMenuItem';
import NumberMenuItem from './menuItems/NumberMenuItem';

interface IAdvancedConfigProps {
    window: Window;
    config: IConfigState;
    updateModelConfig: (payload: PartialConfig) => void;
    updateDetectionConfig: (payload: PartialConfig) => void;
    updateAppConfig: (payload: PartialConfig) => void;
}
export default function AdvancedConfig(props: IAdvancedConfigProps) {
    return (
        <React.Fragment>
            <CheckBoxMenuItem
                window={props.window}
                alert={false}
                name={'Show Reflection'}
                configName={'toggleReflection'}
                helpWith={HelpWith.REFLECTION}
                checked={props.config.toggleReflection}
                onInputChange={props.updateAppConfig}
            />

            {props.config.toggleReflection && (
                <NumberMenuItem
                    name={'Reflect Opacity'}
                    configName={'reflectionOpacity'}
                    step={0.01}
                    defaultValue={props.config.reflectionOpacity}
                    onValidInput={props.updateAppConfig}
                    helpWith={HelpWith.REFLECTION_OPACITY}
                    min={0.01}
                    max={1.0}
                />
            )}
            <CheckBoxMenuItem
                window={props.window}
                alert={false}
                name={'Toggle Debug'}
                configName={'toggleDebug'}
                helpWith={HelpWith.DEBUG}
                checked={props.config.toggleDebug}
                onInputChange={props.updateAppConfig}
            />
            {props.config.toggleDebug && (
                <CanvasMenuItem
                    name={'Camera'}
                    helpWith={HelpWith.VIDEO_STREAM}
                    videoIndex={0}
                />
            )}

            <h3>Model Config</h3>
            <DropDownMenuItem
                name={'Pose Model'}
                configName={'architecture'}
                onInputChange={props.updateModelConfig}
                values={['MobileNetV1', 'ResNet50']}
                defaultValue={props.config.modelConfig.architecture}
                helpWith={HelpWith.ARCHITECTURE}
            />
            <DropDownMenuItem
                name={'Output Stride'}
                configName={'outputStride'}
                onInputChange={props.updateModelConfig}
                values={
                    props.config.modelConfig.architecture === 'MobileNetV1'
                        ? ['8', '16']
                        : ['16', '32']
                }
                defaultValue={props.config.modelConfig.outputStride.toString()}
                helpWith={HelpWith.OUTPUT_STRIDE}
            />
            <DropDownMenuItem
                name={'Resolution'}
                configName={'inputResolution'}
                onInputChange={props.updateModelConfig}
                values={[
                    '161',
                    '193',
                    '257',
                    '289',
                    '321',
                    '353',
                    '385',
                    '417',
                    '449',
                    '481',
                    '513',
                    '801',
                ]}
                defaultValue={props.config.modelConfig.inputResolution.toString()}
                helpWith={HelpWith.RESOLUTION}
            />
            {props.config.modelConfig.architecture === 'MobileNetV1' && (
                <DropDownMenuItem
                    name={'Multiplier'}
                    configName={'multiplier'}
                    onInputChange={props.updateModelConfig}
                    values={['0.50', '0.75', '1']}
                    defaultValue={props.config.modelConfig.multiplier.toString()}
                    helpWith={HelpWith.MULTIPLIER}
                />
            )}

            <h3>Detection Config</h3>
            <NumberMenuItem
                name={'Detections'}
                configName={'maxDetections'}
                step={1}
                defaultValue={props.config.detectionConfig.maxDetections}
                onValidInput={props.updateDetectionConfig}
                helpWith={HelpWith.DETECTIONS}
                min={1}
            />
            <CheckBoxMenuItem
                window={props.window}
                alert={false}
                name={'Flip Horizontal'}
                configName={'flipHorizontal'}
                helpWith={HelpWith.FLIP}
                checked={props.config.detectionConfig.flipHorizontal}
                onInputChange={props.updateDetectionConfig}
            />
            <NumberMenuItem
                name={'Min Score'}
                configName={'scoreThreshold'}
                helpWith={HelpWith.MIN_SCORE}
                step={0.01}
                defaultValue={props.config.detectionConfig.scoreThreshold}
                onValidInput={props.updateDetectionConfig}
                min={0}
                max={1}
            />
            <NumberMenuItem
                name={'NMS Radius'}
                configName={'nmsRadius'}
                helpWith={HelpWith.NMS_RADIUS}
                step={1}
                defaultValue={props.config.detectionConfig.nmsRadius}
                onValidInput={props.updateDetectionConfig}
                min={1}
            />
        </React.Fragment>
    );
}
