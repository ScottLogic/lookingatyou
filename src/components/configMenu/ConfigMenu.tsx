import { Button } from '@material-ui/core';
import React from 'react';
import isEqual from 'react-fast-compare';
import { connect } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { configMenuConsts } from '../../AppConstants';
import {
    resetConfigAction,
    updateConfigAction,
} from '../../store/actions/config/actions';
import {
    ConfigSetAction,
    IAdvancedConfig,
    IAppConfig,
    PartialConfig,
    UpdateConfigAction,
} from '../../store/actions/config/types';
import { IRootStore } from '../../store/reducers/rootReducer';
import {
    getAdvancedConfig,
    getAppConfig,
} from '../../store/selectors/configSelectors';
import AdvancedConfigItems from './AdvancedConfigItems';
import './ConfigMenu.css';
import Help, { HelpWith } from './Help';
import UserConfigItems from './UserConfigItems';

export interface IConfigMenuProps {
    window: Window;
}

interface IConfigMenuMapStateToProps {
    appConfig: IAppConfig;
    advancedConfig: IAdvancedConfig;
}

interface IConfigMenuMapDispatchToProps {
    updateAppConfig: UpdateConfigAction;
    updateModelConfig: UpdateConfigAction;
    updateDetectionConfig: UpdateConfigAction;
    updateAdvancedConfig: UpdateConfigAction;
    resetConfig: () => void;
}

export type ConfigMenuProps = IConfigMenuProps &
    IConfigMenuMapStateToProps &
    IConfigMenuMapDispatchToProps;

interface IConfigMenuState {
    leftPosition: string;
    isUnderMouse: boolean;
}

export class ConfigMenu extends React.Component<
    ConfigMenuProps,
    IConfigMenuState
> {
    private hideTimeout: number = 0;

    constructor(props: ConfigMenuProps) {
        super(props);
        this.state = {
            leftPosition: '0px',
            isUnderMouse: false,
        };
        this.onMouseEnter = this.onMouseEnter.bind(this);
        this.onMouseLeave = this.onMouseLeave.bind(this);
        this.mouseMoveHandler = this.mouseMoveHandler.bind(this);
        this.props.window.addEventListener('mousemove', this.mouseMoveHandler);
    }

    mouseMoveHandler() {
        this.setState({ leftPosition: '0px' });
        this.props.window.clearInterval(this.hideTimeout);
        if (
            !this.state.isUnderMouse &&
            (!this.props.advancedConfig.toggleDebug &&
                this.props.appConfig.toggleAdvanced)
        ) {
            this.hideTimeout = this.props.window.setTimeout(
                () =>
                    this.setState({
                        leftPosition: configMenuConsts.leftPos,
                    }),
                configMenuConsts.visibleTimer,
            );
        }
    }

    onMouseEnter() {
        this.props.window.clearInterval(this.hideTimeout);
        this.setState({ isUnderMouse: true });
    }

    onMouseLeave() {
        this.setState({ isUnderMouse: false });
    }

    shouldComponentUpdate(
        nextProps: ConfigMenuProps,
        nextState: IConfigMenuState,
    ) {
        return (
            nextState.leftPosition !== this.state.leftPosition ||
            !isEqual(nextProps, this.props)
        );
    }

    componentWillUnmount() {
        this.props.window.clearInterval(this.hideTimeout);
        this.props.window.removeEventListener(
            'mousemove',
            this.mouseMoveHandler,
        );
    }

    render() {
        const showAppHelp = () => {
            this.props.updateAppConfig({ showHelp: true });
        };
        return (
            <>
                <div
                    style={{
                        left: this.state.leftPosition,
                        width: configMenuConsts.width,
                    }}
                    className={'ConfigMenu'}
                    onMouseEnter={this.onMouseEnter}
                    onMouseLeave={this.onMouseLeave}
                >
                    <h1>Settings</h1>
                    <Button
                        variant="contained"
                        className="icon"
                        onClick={showAppHelp}
                    >
                        ?
                    </Button>
                    <UserConfigItems {...this.props} />
                    {this.props.appConfig.toggleAdvanced && (
                        <AdvancedConfigItems {...this.props} />
                    )}

                    <br />

                    <Button
                        variant="contained"
                        className="reset"
                        onClick={this.props.resetConfig}
                    >
                        RESET TO DEFAULTS
                    </Button>

                    <br />
                    <br />
                </div>

                {Object.values(HelpWith).map((type, key: number) => (
                    <Help key={key} problemWith={HelpWith[type] as HelpWith} />
                ))}
            </>
        );
    }
}
const mapStateToProps = (state: IRootStore): IConfigMenuMapStateToProps => ({
    appConfig: getAppConfig(state),
    advancedConfig: getAdvancedConfig(state),
});

const mapDispatchToProps = (
    dispatch: ThunkDispatch<IRootStore, void, Action>,
    ownProps: IConfigMenuProps,
) => ({
    updateAppConfig: (payload: PartialConfig) =>
        dispatch(
            updateConfigAction(ConfigSetAction.APP, payload, ownProps.window),
        ),
    updateAdvancedConfig: (payload: PartialConfig) =>
        dispatch(
            updateConfigAction(
                ConfigSetAction.ADVANCED,
                payload,
                ownProps.window,
            ),
        ),
    updateModelConfig: (payload: PartialConfig) =>
        dispatch(
            updateConfigAction(ConfigSetAction.MODEL, payload, ownProps.window),
        ),
    updateDetectionConfig: (payload: PartialConfig) =>
        dispatch(
            updateConfigAction(
                ConfigSetAction.DETECTION,
                payload,
                ownProps.window,
            ),
        ),
    resetConfig: () => dispatch(resetConfigAction(ConfigSetAction.RESET)),
});

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(ConfigMenu);
