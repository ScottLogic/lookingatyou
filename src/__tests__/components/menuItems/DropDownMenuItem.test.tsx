import { mount, shallow } from 'enzyme';
import React from 'react';
import { HelpWith } from '../../../components/configMenu/Help';
import DropDownMenuItem, {
    IDropDownMenuItemProps,
} from '../../../components/configMenu/menuItems/DropDownMenuItem';

let props: IDropDownMenuItemProps;

describe('DropDownMenuItem', () => {
    beforeEach(() => {
        props = {
            name: 'test',
            configName: 'test',
            onInputChange: jest.fn(),
            values: ['a', 'b', 'c'],
            defaultValue: 'a',
            helpWith: HelpWith.OUTPUT_STRIDE,
        };
    });

    it('should render correctly', () => {
        const wrapper = shallow(<DropDownMenuItem {...props} />).debug();
        expect(wrapper).toMatchSnapshot();
    });
});
