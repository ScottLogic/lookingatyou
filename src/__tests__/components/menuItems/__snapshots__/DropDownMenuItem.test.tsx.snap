// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownMenuItem should render correctly 1`] = `
"<div data-tip={true} data-for=\\"OUTPUT_STRIDE\\">
  <label>
    test
  </label>
  <WithStyles(ForwardRef(FormControl)) className=\\"selectForm\\">
    <WithStyles(ForwardRef(Select)) value=\\"a\\" onChange={[Function: handleChange]} name=\\"test\\">
      <WithStyles(ForwardRef(MenuItem)) value=\\"a\\">
        a
      </WithStyles(ForwardRef(MenuItem))>
      <WithStyles(ForwardRef(MenuItem)) value=\\"b\\">
        b
      </WithStyles(ForwardRef(MenuItem))>
      <WithStyles(ForwardRef(MenuItem)) value=\\"c\\">
        c
      </WithStyles(ForwardRef(MenuItem))>
    </WithStyles(ForwardRef(Select))>
  </WithStyles(ForwardRef(FormControl))>
</div>"
`;
